blake3-air/src/columns.rs

```diff
@@ -0,0 +1,114 @@
+use core::borrow::{Borrow, BorrowMut};
+use core::mem::size_of;
+
+use crate::constants::U32_LIMBS;
+
+/// Columns for a Blake-3 AIR which computes one permutation per row.
+///
+/// This is a pretty wide trace but that should be fine.
+#[repr(C)]
+pub struct Blake3Cols<T> {
+    // The inputs to the hash function.
+    pub inputs: [[T; 32]; 16],
+
+    // The chaining values are the first eight outputs of the previous compression.
+    pub chaining_values: [[[T; 32]; 4]; 2],
+
+    // A few auxillary values use to flesh out the first state.
+    pub counter_low: [T; 32],
+    pub counter_hi: [T; 32],
+    pub block_len: [T; 32],
+    pub flags: [T; 32],
+
+    // It should be possible to remove these two but this makes a negligible difference to the overall width of the trace.
+    pub initial_row0: [[T; U32_LIMBS]; 4],
+    pub initial_row2: [[T; U32_LIMBS]; 4],
+
+    pub full_rounds: [FullRound<T>; 7],
+
+    pub final_round_helpers: [[T; 32]; 4],
+
+    pub outputs: [[[T; 32]; 4]; 4],
+}
+
+/// A state at a single instance of time.
+///
+/// Rows `0` and `2` are saved as `2` `16` bit limbs.
+/// Rows `1` and `3` are saved as `32` boolean values.
+#[repr(C)]
+pub struct Blake3State<T> {
+    pub row0: [[T; U32_LIMBS]; 4],
+    pub row1: [[T; 32]; 4],
+    pub row2: [[T; U32_LIMBS]; 4],
+    pub row3: [[T; 32]; 4],
+}
+
+/// Full round columns.
+#[repr(C)]
+pub struct FullRound<T> {
+    // A full round of the Blake3 hash consists of 8 applications of the mixing function.
+    // the first four mixing functions act on the four columns and the second four
+    // functions act on the diagonals.
+    //
+    // We use the output of the previous row to get the input to this row.
+    //
+    /// The outputs after applying the first half of the first 4 mixing functions
+    pub state_prime: Blake3State<T>,
+
+    /// The outputs after applying the first 4 row-mixing functions.
+    pub state_middle: Blake3State<T>,
+
+    /// The outputs after applying the first half of the diagonal mixing functions.
+    pub state_middle_prime: Blake3State<T>,
+
+    /// This will also be the input to the next row.
+    pub state_output: Blake3State<T>,
+}
+
+/// Data needed to verify a single QuarterRound mixing function.
+#[repr(C)]
```

Comment:
Looks like it's not really part of the cols struct that's transmuted, so not strictly needed, but doesn't hurt to keep if you prefer

Reply:
Hm either way seems fine